tools/cli/src/commands/Command.ts

```diff
@@ -0,0 +1,166 @@
+    public readonly helpSections: Section[] = [];
+    private _execute?: (
+        this: this,
+        values: Partial<T>,
+        stack?: Command[],
+    ) => Result<number, string> | Promise<Result<number, string>>;
+
+    constructor(public readonly definition: ExtendedOptionDefinition) {}
+
+    public subcommand(command: Command): this {
+        this.subcommands.push(command);
+        return this;
+    }
+
+    public option<N extends keyof T & string>(option: ExtendedOptionDefinition<N>): this {
+        this.options.push(option);
+        return this;
+    }
+
+    public help(...sections: Section[]): this {
+        this.helpSections.push(...sections);
+        return this;
+    }
+
+    public execute(fn: Command<T>['_execute']): this {
+        this._execute = fn;
+        return this;
+    }
+
+    public async run(argv: string[], stack?: Command[]): Promise<Result<number, CommandError>> {
+        const stackOptions = stack?.flatMap((s) => s.options.filter((o) => o.global)) ?? [];
+        const localOptions = [...this.options, ...stackOptions];
+        const subcommandOption = {
+            name: '_subcommand',
+            defaultOption: true,
+        };
+
+        const subCommandMode = !!this.subcommands.length;
+        if (subCommandMode) {
+            localOptions.push(subcommandOption);
+        }
+
+        const values = commandLineArgs(localOptions, {
+            argv,
+            stopAtFirstUnknown: subCommandMode,
+        });
+
+        if (subCommandMode && values._subcommand) {
+            const subcommand = this.subcommands.find((s) => s.definition.name === values._subcommand);
+            if (subcommand) {
+                const subcommandValues = commandLineArgs([subcommandOption], { argv, stopAtFirstUnknown: true });
+                return await subcommand.run(subcommandValues._unknown ?? [], [...(stack ?? []), this]);
+            }
+        }
+
+        const logger = createLogger(values as LoggerOptions);
+
+        if (values.help) {
+            logger.output(this.getHelpMessage(stack));
+            return Ok(0);
+        }
+
+        if (this._execute) {
+            return (await this._execute.call(this, values as T, stack)).mapErr((error) => ({
+                command: this,
+                error,
+                stack,
+            }));
+        }
+
+        logger.info(this.getHelpMessage(stack));
+
+        if (subCommandMode) {
+            return Err({ command: this, stack, error: 'Unknown command.' });
+        }
+
+        return Err({ command: this, stack, error: 'Unknown option.' });
+    }
+
+    public getHelpMessage(stack?: Command[]) {
+        const globalOptions = [
+            ...this.options.filter((o) => o.global),
+            ...(stack?.flatMap((o) => o.options.filter((o) => o.global)) ?? []),
+        ];
+
+        const nonGlobalOptions = this.options.filter((o) => !o.global);
+
+        let cmd = 'backtrace';
```

Comment:
can we expose prefix to const? I think we want to have it in const in case we want to change it for a different command. `Backtrace` sounds weird because you won't use it to other operations you can do in Backtrace.

Reply:
Done.